src/table/io/FitsSchemaInputMapper.cc

```diff
@@ -0,0 +1,619 @@
+        item.*Member = _v;
+    }
+    explicit SetFitsSchemaString(std::string const & v) : _v(v) {}
+private:
+    std::string const & _v;
+};
+
+} // anonymous
+
+class FitsSchemaInputMapper::Impl {
+public:
+
+    // A container class (based on Boost.MultiIndex) that provides three sort orders,
+    // on column number, flag bit, and name (ttype).  This allows us to insert fields into the
+    // schema in the correct order, regardless of which order they appear in the
+    // FITS header.
+    typedef boost::multi_index_container<
+        FitsSchemaItem,
+        boost::multi_index::indexed_by<
+            boost::multi_index::ordered_non_unique<
+                boost::multi_index::member<FitsSchemaItem,int,&FitsSchemaItem::column>
+            >,
+            boost::multi_index::ordered_non_unique<
+                boost::multi_index::member<FitsSchemaItem,int,&FitsSchemaItem::bit>
+            >,
+            boost::multi_index::hashed_unique<
+                boost::multi_index::member<FitsSchemaItem,std::string,&FitsSchemaItem::ttype>
+            >,
+            boost::multi_index::sequenced<>
+        >
+    > InputContainer;
+
+    // Typedefs for the special functors used to set data members.
+    typedef SetFitsSchemaString<&FitsSchemaItem::ttype> SetTTYPE;
+    typedef SetFitsSchemaString<&FitsSchemaItem::tform> SetTFORM;
+    typedef SetFitsSchemaString<&FitsSchemaItem::tccls> SetTCCLS;
+    typedef SetFitsSchemaString<&FitsSchemaItem::tunit> SetTUNIT;
+    typedef SetFitsSchemaString<&FitsSchemaItem::doc> SetDoc;
+
+    // Typedefs for the different indices.
+    typedef InputContainer::nth_index<0>::type ByColumn;
+    typedef InputContainer::nth_index<1>::type ByBit;
+    typedef InputContainer::nth_index<2>::type ByName;
+    typedef InputContainer::nth_index<3>::type AsList;
+
+    // Getters for the different indices.
+    ByColumn & byColumn() { return inputs.get<0>(); }
+    ByBit & byBit() { return inputs.get<1>(); }
+    ByName & byName() { return inputs.get<2>(); }
+    AsList & asList() { return inputs.get<3>(); }
+
+    Impl() : version(0), flagColumn(0), archiveHdu(-1) {}
+
+    int version;
+    int flagColumn;
+    int archiveHdu;
+    Schema schema;
+    std::vector<std::unique_ptr<FitsColumnReader>> readers;
+    std::vector<Key<Flag>> flagKeys;
+    std::unique_ptr<bool[]> flagWorkspace;
+    PTR(io::InputArchive) archive;
+    InputContainer inputs;
+};
+
+FitsSchemaInputMapper::FitsSchemaInputMapper(daf::base::PropertyList & metadata, bool stripMetadata) :
+    _impl(boost::make_shared<Impl>())
+{
+    // Set the table version.  If AFW_TABLE_VERSION tag exists, use that
+    // If not, set to 0 if it has an AFW_TYPE, Schema default otherwise (DM-590)
+    if (!metadata.exists("AFW_TYPE")) {
+        _impl->version = lsst::afw::table::Schema::VERSION;
+    }
+    _impl->version = metadata.get("AFW_TABLE_VERSION", _impl->version);
+    if (stripMetadata) {
+        metadata.remove("AFW_TABLE_VERSION");
```

Comment:
Good point.  I had it in my head that I couldn't remove it here because I needed it to still be present in some downstream processing, but looking again I must have reordered things so that's no longer the case.